File: app/src/main/java/com/ivanovsky/passnotes/util/InputOutputUtils.java
package com.ivanovsky.passnotes.util;

import androidx.annotation.Nullable;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.util.concurrent.atomic.AtomicBoolean;

import timber.log.Timber;

public class InputOutputUtils {

    private static final int BUFFER_SIZE = 1024 * 8;

    public static FileInputStream newFileInputStreamOrNull(File file) {
        FileInputStream result = null;

        try {
            result = new FileInputStream(file);
        } catch (FileNotFoundException e) {
            Timber.d(e);
        }

        return result;
    }

    public static FileOutputStream newFileOutputStreamOrNull(File file) {
        FileOutputStream result = null;

        try {
            result = new FileOutputStream(file);
        } catch (FileNotFoundException e) {
            Timber.d(e);
        }

        return result;
    }

    public static String toString(InputStream is) {
        String result = null;

        BufferedReader reader = null;
        try {
            reader = new BufferedReader(new InputStreamReader(is, "UTF-8"));
            StringBuilder sb = new StringBuilder();
            String line;

            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }

            result = sb.toString();
        } catch (IOException e) {
            Timber.d(e);
        } finally {
            if (reader != null) {
                try {
                    reader.close();
                } catch (IOException e) {
                    Timber.d(e);
                }
            }
        }

        return result;
    }

    public static void copy(InputStream in, OutputStream out, boolean close) throws IOException {
        byte[] buf = new byte[BUFFER_SIZE];
        int len;

        try {
            while ((len = in.read(buf)) > 0) {
                out.write(buf, 0, len);
            }

            out.flush();
        } finally {
            if (close) {
                try {
                    in.close();
                } catch (IOException e) {
                    Timber.d(e);
                }

                try {
                    out.close();
                } catch (IOException e) {
                    Timber.d(e);
                }
            }
        }
    }

    public static void copy(InputStream in, OutputStream out, boolean close,
                            AtomicBoolean cancellation) throws IOException {
        byte[] buf = new byte[BUFFER_SIZE];
        int len;

        while ((len = in.read(buf)) > 0 && !cancellation.get()) {
            out.write(buf, 0, len);
        }

        out.flush();

        if (close) {
            in.close();
            out.close();
        }
    }

    public static void close(@Nullable OutputStream out) {
        if (out == null) {
            return;
        }

        try {
            out.close();
        } catch (IOException e) {
            Timber.d(e);
        }
    }

    public static void close(@Nullable InputStream in) {
        if (in == null) {
            return;
        }

        try {
            in.close();
        } catch (IOException e) {
            Timber.d(e);
        }
    }

    private InputOutputUtils() {
    }
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/util/InputOutputUtils.kt
package com.ivanovsky.passnotes.util

import com.ivanovsky.passnotes.data.entity.OperationError.newGenericIOError
import com.ivanovsky.passnotes.data.entity.OperationResult
import timber.log.Timber
import java.io.Closeable
import kotlin.Throws
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.concurrent.atomic.AtomicBoolean

object InputOutputUtils {

    private const val BUFFER_SIZE = 1024 * 8
    private val UNCANCELABLE = AtomicBoolean(false)

    @JvmStatic
    fun newFileInputStreamOrNull(file: File): FileInputStream? {
        return try {
            FileInputStream(file)
        } catch (e: FileNotFoundException) {
            Timber.d(e)
            null
        }
    }

    @JvmStatic
    fun newFileOutputStreamOrNull(file: File): FileOutputStream? {
        return try {
            FileOutputStream(file)
        } catch (e: FileNotFoundException) {
            Timber.d(e)
            null
        }
    }

    @JvmStatic
    fun copy(
        sourceFile: File,
        destinationFile: File
    ): OperationResult<Unit> {
        val source = try {
            FileInputStream(sourceFile)
        } catch (exception: FileNotFoundException) {
            Timber.d(exception)
            return OperationResult.error(newGenericIOError(exception))
        }

        val destination = try {
            FileOutputStream(destinationFile, false)
        } catch (exception: FileNotFoundException) {
            Timber.d(exception)
            return OperationResult.error(newGenericIOError(exception))
        }

        return try {
            copyOrThrow(
                source,
                destination,
                isCloneOnFinish = true,
                cancellation = UNCANCELABLE
            )
            OperationResult.success(Unit)
        } catch (exception: IOException) {
            Timber.d(exception)
            OperationResult.error(newGenericIOError(exception))
        }
    }

    @JvmStatic
    fun copy(
        source: InputStream,
        destinationFile: File
    ): OperationResult<Unit> {
        val destination = try {
            FileOutputStream(destinationFile, false)
        } catch (exception: FileNotFoundException) {
            Timber.d(exception)
            return OperationResult.error(newGenericIOError(exception))
        }

        return try {
            copyOrThrow(
                source,
                destination,
                isCloneOnFinish = true,
                cancellation = UNCANCELABLE
            )
            OperationResult.success(Unit)
        } catch (exception: IOException) {
            Timber.d(exception)
            OperationResult.error(newGenericIOError(exception))
        }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun copyOrThrow(
        source: InputStream,
        destination: OutputStream,
        isCloneOnFinish: Boolean
    ) {
        copyOrThrow(source, destination, isCloneOnFinish, cancellation = UNCANCELABLE)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun copyOrThrow(
        source: InputStream,
        destination: OutputStream,
        isCloneOnFinish: Boolean,
        cancellation: AtomicBoolean
    ) {
        try {
            val buf = ByteArray(BUFFER_SIZE)
            var len: Int
            while (source.read(buf).also { len = it } > 0 && !cancellation.get()) {
                destination.write(buf, 0, len)
            }
            destination.flush()
        } finally {
            if (isCloneOnFinish) {
                closeOrThrow(source)
                closeOrThrow(destination)
            }
        }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun closeOrThrow(closeable: Closeable?) {
        if (closeable == null) {
            return
        }

        closeable.close()
    }

    @JvmStatic
    fun close(closeable: Closeable?) {
        if (closeable == null) {
            return
        }

        try {
            closeable.close()
        } catch (e: IOException) {
            Timber.d(e)
        }
    }

    @JvmStatic
    fun readAllBytes(
        source: InputStream,
        isCloseOnFinish: Boolean
    ): OperationResult<ByteArray> {
        return try {
            val bytes = source.readBytes()

            if (isCloseOnFinish) {
                close(source)
            }

            OperationResult.success(bytes)
        } catch (e: IOException) {
            Timber.e(e)
            OperationResult.error(newGenericIOError(e))
        }
    }
}